torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -209,9 +214,23 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
           return;
         }
 
-        // Accumulate all the gradients in the context.
         TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
+
+        // Accumulate all the gradients in the context.
         for (size_t i = 0; i < grads.size(); i++) {
+          // Compute the number of references held to the same grad tensor that
+          // have already been processed. This ensures that if have multiple
+          // references to the same grad tensor in our grads list, all but one
+          // tensor would be cloned when we accumulate grad. This avoids a
+          // clone for the last reference of the grad in our grads list.
+          // See use_count check in AccumulateGrad::accumulateGradAndCallHooks.
+          size_t num_expected_refs = 1;
+          if (i != 0) {
+            for (int64_t j = i - 1; j >= 0; j--) {
+              num_expected_refs += (grads[j].is_same(grads[i]));
+            }
+          }
```

Comment:
I'd suggest leveraging `unsafeGetTensorImpl()` and a hash table to avoid the quadratic time complexity here.

```
std::unordered_map<const TensorImpl*, int> tensor_to_refs;
for (size_t i = 0; i < grads.size(); i++) {
  const auto num_expected_refs = (tensor_to_refs[grads[i].unsafeGetTensorImpl()] += 1);

```

Note that `is_same()` is implemented as.

https://github.com/pytorch/pytorch/blob/f9f135c5d8fc1986e112e62473b3498cdf4d1c43/aten/src/ATen/templates/TensorBody.h#L174-L176

Reply:
@pritamdamania87  and I discussed this and we're just going to delete this "optimization" as it is incredibly unlikely to happen in practice